src/Compression/CompressionCodecSZ3.cpp

```diff
@@ -0,0 +1,307 @@
+
+protected:
+    bool isCompression() const override { return true; }
+    bool isGenericCompression() const override { return true; }
+    /// SZ3 is still under development, it writes its current version into the serialized compressed data.
+    /// Therefore, update SZ3 with care to avoid breaking existing persistencies.
+    /// We mark it as experimental for now.
+    bool isExperimental() const override { return true; }
+    bool isVectorCodec() const override { return true; }
+    String getDescription() const override { return "SZ3 is a lossy compressor for floating-point data with error bounds."; }
+
+private:
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override;
+    void doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const override;
+
+    UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override;
+
+    std::vector<size_t> dimensions;
+    const UInt8 float_width;
+    const SZ3::ALGO algorithm;
+    const SZ3::EB error_bound_mode;
+    const Float64 error_value;
+};
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int CORRUPTED_DATA;
+    extern const int ILLEGAL_CODEC_PARAMETER;
+    extern const int LOGICAL_ERROR;
+}
+
+String getSZ3AlgorithmString(SZ3::ALGO algorithm)
+{
+    for (const auto & [algorithm_string, algorithm_id] : SZ3::ALGO_MAP)
+    {
+        if (algorithm_id == algorithm)
+            return algorithm_string;
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid algorithm");
+}
+
+String getSZ3ErrorBoundModeString(SZ3::EB error_bound_mode)
+{
+    for (const auto & [error_bound_string, error_bound_mode_id] : SZ3::EB_MAP)
+    {
+        if (error_bound_mode_id == error_bound_mode)
+            return error_bound_string;
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid error bound mode");
+}
+
+CompressionCodecSZ3::CompressionCodecSZ3(
+    UInt8 float_size_, SZ3::ALGO algorithm_, SZ3::EB error_bound_mode_, double error_value_)
+    : float_width(float_size_)
+    , algorithm(algorithm_)
+    , error_bound_mode(error_bound_mode_)
+    , error_value(error_value_)
+{
+    setCodecDescription(
+        "SZ3", {std::make_shared<ASTLiteral>(getSZ3AlgorithmString(algorithm)),
+                std::make_shared<ASTLiteral>(getSZ3ErrorBoundModeString(error_bound_mode)),
+                std::make_shared<ASTLiteral>(error_value)});
+}
+
+uint8_t CompressionCodecSZ3::getMethodByte() const
+{
+    return static_cast<uint8_t>(CompressionMethodByte::SZ3);
+}
+
+void CompressionCodecSZ3::updateHash(SipHash & hash) const
+{
+    getCodecDesc()->updateTreeHash(hash, true);
+    hash.update(float_width);
+}
+
+UInt32 CompressionCodecSZ3::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    return sizeof(UInt8) + uncompressed_size;
+}
+
+UInt32 CompressionCodecSZ3::doCompressData(const char * source, UInt32 source_size, char * dest) const
```

Comment:
I did this:

```sql
CREATE TABKE tab (key UInt64, val Array(Float64) CODEC(SZ3('ALGO_INTERP', 'REL', 0.01))) ENGINE = MergeTree ORDER BY key;
INSERT INTO tab VALUES (1, [1.0, 2.0, 3.0]);
```

and I see that "::doCompressData" is called twice. I guess this is because the array is encoded as two streams ... one for the offsets and one for the values. So it would SZ3-encode both ... it seems wrong to compress array offsets with a lossy algorithm.